src/main/java/com/aws/greengrass/clientdevices/auth/CertificateManager.java

```diff
@@ -167,100 +171,102 @@ public String getCaPassPhrase() {
+                    CertificateUpdateEvent certificateUpdateEvent =
+                            new CertificateUpdateEvent(keyPair, cert, caCertificates);
+                    request.getCertificateUpdateConsumer().accept(certificateUpdateEvent);
+                };
+                if (type.equals(GetCertificateRequestOptions.CertificateType.SERVER)) {
+                    return new ServerCertificateGenerator(CertificateHelper.getX500Name(request.getServiceName()),
+                            keyPair.getPublic(), consumer, certificateStore, certificatesConfig, clock);
+                } else {
+                    return new ClientCertificateGenerator(CertificateHelper.getX500Name(request.getServiceName()),
+                            keyPair.getPublic(), consumer, certificateStore, certificatesConfig, clock);
+                }
+            case CUSTOM:
+                // the customer certificate generator is responsible for getting the certificate store and clock
+                if (request instanceof CustomGeneratorCertificateRequest) {
+                    return ((CustomGeneratorCertificateRequest) request).getCertificateGenerator();
+                }
+                throw new CertificateGenerationException("Invalid custom certificate request");
+            default:
+                throw new CertificateGenerationException("Unsupported certificate type: " + type);
         }
     }
 
-    private void subscribeToServerCertificateUpdatesNoCSR(@NonNull GetCertificateRequest certificateRequest,
-                                                          @NonNull PublicKey publicKey,
-                                                          @NonNull BiConsumer<X509Certificate, X509Certificate[]> cb)
+    private void subscribeToCertificateUpdatesNoCSR(@NonNull CertificateGenerator certificateGenerator,
+                                                    @NonNull GetCertificateRequest certificateRequest,
+                                                    @NonNull GetCertificateRequestOptions.CertificateType type)
             throws CertificateGenerationException {
-        CertificateGenerator certificateGenerator =
-                new ServerCertificateGenerator(CertificateHelper.getX500Name(certificateRequest.getServiceName()),
-                        publicKey, cb, certificateStore, certificatesConfig, clock);
-
-        // Add certificate generator to monitors first in order to avoid missing events
-        // that happen while the initial certificate is being generated.
+        // Add certificate generator to monitors
         certExpiryMonitor.addToMonitor(certificateGenerator);
-        cisShadowMonitor.addToMonitor(certificateGenerator);
         caConfigurationMonitor.addToMonitor(certificateGenerator);
 
+        if (type == GetCertificateRequestOptions.CertificateType.SERVER
+                || type == GetCertificateRequestOptions.CertificateType.CUSTOM) {
```

Comment:
not sure about adding a custom type. Really what you want is client+server right? I think that adding that would be more clear in that it is the combination of both of what we have today.

Reply:
yes I can rename to ClientAndServer. Passing a `CertificateGenerator` in the request also allows the caller to controll the expiry time, SANs, and other configs, so I think that bit of refactoring is still needed.

Reply:
Do we need a new type at all? Would it make sense to modify ClientDeviceAuthApi to accept a list of cert types in the request and all the other customizations (SAN, expiry, etc)?

Reply:
I'm trying to keep backwards compatibility so I do not want to change anything in the original `GetCertificateRequest`.

Conceptually, I think today there is an 1:1 mapping between cert type and certificate. Both `CertificateSubscriptionEvent` and `GetCertificateRequestOptions` relies on singleton cert type, which is why I prefer to not use a list of cert types.

Reply:
For my understanding how does this break compatibility if we just modify the api class but not the ipc handler?

Reply:
If we modify the API class input type, old versions of other components (e.g. moquette server [here](https://github.com/aws-greengrass/aws-greengrass-moquette-mqtt/blob/10d8671b4622a1c388c0f2ad43f3a3f521e084e9/integration/src/main/java/com/aws/greengrass/mqtt/moquette/MQTTService.java#L106)) would not be able to call `subscribeToCertificateUpdates`.

This can be addressed by bumping dependency version range on CDA but I would like to avoid it unless necessary.